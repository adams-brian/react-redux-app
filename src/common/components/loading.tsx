import * as React from 'react';

interface IState {
  timeoutFired: boolean;
}

export interface IStateProps {
  error: string;
  loaded: boolean;
}
export interface IDispatchProps {
  load: () => void;
}

export default <TProps extends {}>(Component: React.ComponentType<TProps>) =>

  class Loading extends React.Component<IStateProps & IDispatchProps & TProps, IState> {
    public static displayName = `loading(${Component.displayName || Component.name})`;
    private timeout: NodeJS.Timer;

    constructor(props: IStateProps & IDispatchProps & TProps) {
        super(props);
        this.state = {
            timeoutFired: false
        };
    }

    public componentDidMount() {
      this.timeout = setTimeout(() => this.setState({timeoutFired: true}), 500);
      if (!this.props.loaded) {
        this.props.load();
      }
    }

    public componentWillUnmount() {
      clearTimeout(this.timeout);
    }

    public render() {
      if (this.props.loaded) {
        return (
          <Component {...this.props} />
        );
      }
      else if (this.props.error && this.props.error.length > 0) {
        return (
          <div>Error: {this.props.error}</div>
        )
      }
      else if (this.state.timeoutFired) {
        return (
          <div>Loading...</div>
        )
      }
      else {
        return null;
      }
    }
  }
