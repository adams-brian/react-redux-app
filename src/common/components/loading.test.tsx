import { mount } from 'enzyme';
import * as React from 'react';
import * as sinon from 'sinon';

import loading from './loading';

const SimpleComponent = (props: { message: string }) =>
  <div>{props.message}</div>

const Wrapper = loading(SimpleComponent);

describe("loading", () => {

  let clock: sinon.SinonFakeTimers;
  let load: sinon.SinonSpy;

  const getElement = (loaded: boolean) => 
    <Wrapper
      error=""
      loaded={loaded}
      load={load}
      message="message from props"
    />;

  beforeEach(() => {
    clock = sinon.useFakeTimers();
    load = sinon.spy();  
  });

  afterEach(() => {
    clock.restore();
  });

  it('renders as expected', () => {
    const component = mount(getElement(false));
    expect(component).toMatchSnapshot();
  });

  it('should render the component if loaded', () => {
    const component = mount(getElement(true));
    expect(component.html()).toMatchSnapshot();
    clock.tick(500);
    expect(component.html()).toMatchSnapshot();
    clock.tick(5000);
    expect(component.html()).toMatchSnapshot();
    expect(load.called).toBe(false);
  });

  it('should render null until loaded for load times < 500ms', () => {
    const component = mount(getElement(false));
    expect(load.calledOnce).toBe(true);
    setTimeout(() => { component.setProps( { loaded: true }); }, 500);
    expect(component.html()).toMatchSnapshot();
    clock.tick(500);
    expect(component.html()).toMatchSnapshot();
    clock.tick(5000);
    expect(component.html()).toMatchSnapshot();
  });

  it('should render null for 500ms, then loading until loaded for load times > 500ms', () => {
    const component = mount(getElement(false));
    expect(load.calledOnce).toBe(true);
    setTimeout(() => { component.setProps( { loaded: true }); }, 2000);
    expect(component.html()).toMatchSnapshot();
    clock.tick(500);
    expect(component.html()).toMatchSnapshot();
    clock.tick(1500);
    expect(component.html()).toMatchSnapshot();
  });

  it('should render the error message', () => {
    const component = mount(getElement(false));
    expect(load.calledOnce).toBe(true);
    setTimeout(() => { component.setProps( { error: 'error message' }); }, 2000);
    expect(component.html()).toMatchSnapshot();
    clock.tick(500);
    expect(component.html()).toMatchSnapshot();
    clock.tick(1500);
    expect(component.html()).toMatchSnapshot();
  });

  it('should clear the timeout on unmount', () => {
    const component = mount(getElement(false));
    expect(component.html()).toMatchSnapshot();
    component.unmount();
    clock.tick(5000);
    expect(component.html()).toMatchSnapshot();
  });

});
