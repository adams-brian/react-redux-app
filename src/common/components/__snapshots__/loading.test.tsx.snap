// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading renders as expected 1`] = `
<loading(SimpleComponent)
  error=""
  load={[Function]}
  loaded={false}
  message="message from props"
/>
`;

exports[`loading should clear the timeout on unmount 1`] = `null`;

exports[`loading should clear the timeout on unmount 2`] = `null`;

exports[`loading should render null for 500ms, then loading until loaded for load times > 500ms 1`] = `null`;

exports[`loading should render null for 500ms, then loading until loaded for load times > 500ms 2`] = `"<div>Loading...</div>"`;

exports[`loading should render null for 500ms, then loading until loaded for load times > 500ms 3`] = `"<div>message from props</div>"`;

exports[`loading should render null until loaded for load times < 500ms 1`] = `null`;

exports[`loading should render null until loaded for load times < 500ms 2`] = `"<div>message from props</div>"`;

exports[`loading should render null until loaded for load times < 500ms 3`] = `"<div>message from props</div>"`;

exports[`loading should render the component if loaded 1`] = `"<div>message from props</div>"`;

exports[`loading should render the component if loaded 2`] = `"<div>message from props</div>"`;

exports[`loading should render the component if loaded 3`] = `"<div>message from props</div>"`;

exports[`loading should render the error message 1`] = `null`;

exports[`loading should render the error message 2`] = `"<div>Loading...</div>"`;

exports[`loading should render the error message 3`] = `"<div>Error: error message</div>"`;
